include/ck/utility/data_type.hpp

```diff
@@ -942,125 +942,160 @@ using int8x16_t = typename vector_type<int8_t, 16>::type;
-{
-    union
+    // convert bfp16 to fp32
+    template <>
+    inline __host__ __device__ constexpr float convert<float, bhalf_t>(bhalf_t x)
     {
-        float fp32;
-        uint32_t int32;
-    } u = {x};
-
-    // When the exponent bits are not all 1s, then the value is zero, normal,
-    // or subnormal. We round the bfloat16 mantissa up by adding 0x7FFF, plus
-    // 1 if the least significant bit of the bfloat16 mantissa is 1 (odd).
-    // This causes the bfloat16's mantissa to be incremented by 1 if the 16
-    // least significant bits of the float mantissa are greater than 0x8000,
-    // or if they are equal to 0x8000 and the least significant bit of the
-    // bfloat16 mantissa is 1 (odd). This causes it to be rounded to even when
-    // the lower 16 bits are exactly 0x8000. If the bfloat16 mantissa already
-    // has the value 0x7f, then incrementing it causes it to become 0x00 and
-    // the exponent is incremented by one, which is the next higher FP value
-    // to the unrounded bfloat16 value. When the bfloat16 value is subnormal
-    // with an exponent of 0x00 and a mantissa of 0x7f, it may be rounded up
-    // to a normal value with an exponent of 0x01 and a mantissa of 0x00.
-    // When the bfloat16 value has an exponent of 0xFE and a mantissa of 0x7F,
-    // incrementing it causes it to become an exponent of 0xFF and a mantissa
-    // of 0x00, which is Inf, the next higher value to the unrounded value.
-    bool flag0 = ~u.int32 & 0x7f800000;
-
-    // When all of the exponent bits are 1, the value is Inf or NaN.
-    // Inf is indicated by a zero mantissa. NaN is indicated by any nonzero
-    // mantissa bit. Quiet NaN is indicated by the most significant mantissa
-    // bit being 1. Signaling NaN is indicated by the most significant
-    // mantissa bit being 0 but some other bit(s) being 1. If any of the
-    // lower 16 bits of the mantissa are 1, we set the least significant bit
-    // of the bfloat16 mantissa, in order to preserve signaling NaN in case
-    // the bfloat16's mantissa bits are all 0.
-    bool flag1 = !flag0 && (u.int32 & 0xffff);
-
-    u.int32 += flag0 ? 0x7fff + ((u.int32 >> 16) & 1) : 0; // Round to nearest, round to even
-    u.int32 |= flag1 ? 0x10000 : 0x0;                      // Preserve signaling NaN
-
-    return uint16_t(u.int32 >> 16);
-}
+        union
+        {
+            uint32_t int32;
+            float fp32;
+        } u = {uint32_t(x) << 16};
 
-// convert bfp16 to fp16 via fp32
-template <>
-inline __host__ __device__ constexpr half_t type_convert<half_t, bhalf_t>(bhalf_t x)
-{
-    float x_fp32 = type_convert<float>(x);
+        return u.fp32;
+    }
 
-    return static_cast<half_t>(x_fp32);
-}
+    // convert fp32 to bfp16
+    template <>
+    inline __host__ __device__ constexpr bhalf_t convert<bhalf_t, float>(float x)
+    {
+        // if using rtn
+        if(BF16ConvertRTN_)
```

Comment:
I think the `BF16ConvertRTN_` should be a constant variable, which is known in compile team. So that the if-statement here can be `if constexpr (BF16ConvertRTN_)` to avoid branch in the kernel.

Reply:
At the same time we want to be able to switch between RTN and RTZ within one kernel, right?

Reply:
Maybe we can do something like this:
```
template <typename Y, typename X, typename config = void>
struct TypeConvert
{
    __host__ __device__ Y operator()(X& x) const
    {
        return static_cast<Y>(x);
    }
};

template <>
struct TypeConvert<bhalf_t, float, integral_constant<bool, true>>
{
    __host__ __device__ bhalf_t operator()(float& x) const
    {
        //RTZ
    }
};


template <>
struct TypeConvert<bhalf_t, float, integral_constant<bool, false>>
{
    __host__ __device__ bhalf_t operator()(float& x) const
    {
        //RTN
    }
};
```
@asroy Do you have any idea?

Reply:
I see your point, indeed, we could check this condition in compile time, but would need separate objects for RTN and RTZ conversion, looks like a plan, thanks!

Reply:
@geyyer Updated a little. Use `Variadic templates`, so we can have more flexibility. For example, for `fp8_t ` type, we may have more than one switches.

```
template <typename Y, typename X, typename... config>
struct TypeConvert
{
    __host__ __device__ Y operator()(X& x) const
    {
        return static_cast<Y>(x);
    }
};

template <>
struct TypeConvert<bhalf_t, float, integral_constant<bool, true>>
{
    __host__ __device__ bhalf_t operator()(float& x) const
    {
        //RTZ
    }
};


template <>
struct TypeConvert<bhalf_t, float, integral_constant<bool, false>>
{
    __host__ __device__ bhalf_t operator()(float& x) const
    {
        //RTN
    }
};


template <>
struct TypeConvert<fp8_t, float, integral_constant<bool, false>,  integral_constant<bool, true>>
{
    __host__ __device__ bhalf_t operator()(float& x) const
    {
        //RTN
    }
};
```